svm-transaction/src/svm_transaction/reference.rs

```diff
@@ -0,0 +1,28 @@
+use {
+    crate::SVMTransaction,
+    solana_sdk::{hash::Hash, signature::Signature, transaction::VersionedTransaction},
+};
+
+// For any type that implements `TransactionTrait`, a reference to that type
+// should also implement `TransactionTrait`.
```

Comment:
```suggestion
// For any type that implements `SVMTransaction`, a reference to that type
// should also implement `SVMTransaction`.
```

Reply:
Is there a particular reason you need this blanket implementation at all? Rust natively applies traits to references.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a715636d1106215a2b9d130651a162a6

Reply:
`T`, `&T` and `&mut T` are different types in Rust, and all need a different trait implementation. I've run into this problem before. See: https://stackoverflow.com/questions/44928882/why-do-i-get-the-error-the-trait-foo-is-not-implemented-for-mut-t-even-th

> T, &T, and &mut T are all different types; and that means that &mut &mut T is likewise a different type. Traits are not automatically implemented for references to a type. If you wish to implement a trait for either of the references, you need to write it out explicitly.

Reply:
Hm, seems to work just fine in my example. Where might you run into this?

Reply:
Try this example:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=292551dcf26efb21b952e829c18e8aca

Reply:
My specific case is that we sometimes have a slice of references `&[&T]` and sometimes have slices of owned `&[T]`.
If `&T` doesn't also implement, then (afaik) `&[&T]` cannot be passed to the function.

Specific example from `Bank`:
```rust
    fn check_status_cache(
        &self,
        sanitized_txs: &[impl core::borrow::Borrow<SanitizedTransaction>],
        lock_results: Vec<TransactionCheckResult>,
        error_counters: &mut TransactionErrorMetrics,
    ) -> Vec<TransactionCheckResult>;
```

this fn is called with both refs and owned SanitizedTransactions and I was able to make that happen using the Borrow trait. Could probably do something similar, but it seems much simpler now that we have a trait abstraction to just say "yeah, a reference to a transaction is able to access transaction data"

Reply:
minimal example from playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=46b8cb62413db10032a1243914c80ac1

Reply:
Ahh I wouldn't expect that example to work. Not because `&Struct(42)` doesn't implement `Trait`, but because the `runner` function specifically asks for an owned object `T`.

I've typically steered away from blanket impl on `&T` to allow functions to require `T` and accept either `Struct` or `&Struct`.

Reply:
> the runner function specifically asks for an owned object T

`T` means any type, including references. It does not mean an owned object.

Reply:
> Change runner to accept &T and you'll get "try borrowing here". This would be my preferred behavior, tbh.

That works fine for fns that take a single `T` or `&T`, but not when we're dealing with slices.
If we want to allow for slices of owned or referenced, we need some sort of generic code; I could definitely get away with implementing those functions using something like:
```rust
    fn check_status_cache<Tx: SVMTransaction>(
        &self,
        txs: &[impl core::borrow::Borrow<Tx>],
        lock_results: Vec<TransactionCheckResult>,
        error_counters: &mut TransactionErrorMetrics,
    ) -> Vec<TransactionCheckResult>;
```

not opposed to doing this since there's only a few that actually require it to be generic in this way.
That's actually probably better than a blanket ref impls, that could lead to some nastily nested code if we have somehow have a `&&T`

Reply:
Removed the blanket implementations: [80c046d](https://github.com/anza-xyz/agave/pull/1918/commits/80c046d3d8335e14b5e01fc9335031a7b34714ce)